### ISSUE TITLE
Provider Review

### ISSUE THREAD

**Issue by catsby** [thetonymaster/terraform-provider-cloudscale#12]
Hello! 

My name is Clint, I'm a member of the Terraform team @ HashiCorp.

I’ve taken a look at the provider here and would like to say great work so far! I do have feedback outlined below that I’d like to see addressed before we move on to the next steps. I believe the issues are all fairly minor and can likely be cleaned up pretty quickly. I’m opening this issues as a sort of checklist for tracking remaining items and discussion. The review was done on the git commit c950dddafb1baa9b3926de1986809c6238562bbb .

- [x] ~resource_floating_ip: `prefix_length` is documented to only accept a value of `56`, and if I read correctly, only when `ip_version` is `6`. If this is the case then consider adding validation functions to their schema definitions. For `ip_version`, validate it is either `4` or `6`. For `prefix_length` unfortunately we can’t validate it is `56` only when `ip_version` is `6`, but you can validate it’s value~ discussed 👍 
- [ ] I believe the markdown documentation files need to have extention `.html.markdown` (or `.html.md`), otherwise the conversion will not have the `.html` extension. Terraform.io is a static html site so that will be required. 
- [x] when deleting a resource, please check the error code. If your API returns a `404` error if the resource is not found (or already deleted) then we need to capture that error, log that the resource could not be found, and call `d.SetId(“”)`. The Server resource does this. See also Route53 Record example: https://github.com/terraform-providers/terraform-provider-aws/blob/master/aws/resource_aws_route53_record.go#L637-L650
- [x] ~resource_server: `volume_size_gb` is said to default to `10`. We should include that default in the schema definition. Right now it’s defined as `Optional` and `ForceNew` . If not a default locally, should it be `Computed` as well? I would need to test it out and view tests to confirm~ discussed 👍 
- [x] resource_server: `ssh_keys` are not read from the API. Are they available? Can you describe the expected behavior of this attribute? 
- [x] resource_server: attributes that are `TypeList`: Lists are generally ordered. If the API does not return them in the same order they are provided, there can be conflict. Consider `TypeSet` for attributes were there could be more than 1 stanza block of it, and ordering is not relevant. If ordered is relevant, we need to ensure Terraform is enforcing the ordering from the configuration.
- [x] resource_server: line 294, 322, 325, 327, etc: when calling `d.Set` for anything that is not a scalar value, it’s important to check the error returned. If the value is not in the correct format, you’ll ignore the error, and the state will match the configuration and not the returned value from the API, which can lead to undetected drift. Here’s an example: https://github.com/terraform-providers/terraform-provider-aws/blob/master/aws/resource_aws_db_instance.go#L825-L831

In summary, things look really good. I have not had a chance to evaluate the tests because I don’t yet have credentials to a test account. I believe someone at HashiCorp will be reaching out to get those.

Let me know if you have any questions! 

Cheers,
Clint

--------

Update from October 18, 2017: 

- Considering all above items addressed when #13 is merged

**Comment by davidhalter** [comment]
Hi Clint, just regarding the documented limits in our API, don't you think it's better that our API side checks this? I think hard coding this doesn't make a lot of sense in the code since we need to validate that anyway.

~ Dave

**Comment by thetonymaster** [comment]
@davidhalter 
I believe this could be an optimization of sorts? So it does not waste time creating other resources? But yeah, I agree on your point.

@catsby: `volume_size_gb` is defaulted by the API in case it is omitted. 
On the ssh keys case, there are three attributes, `ssh_keys` to provide keys, and `ssh_fingerprints`, ` ssh_host_keys` are used to retrieve the keys on a server.

**Comment by davidhalter** [comment]
> I believe this could be an optimization of sorts? So it does not waste time creating other resources?

Well I just think that those "optimizations" are not really worth it, because it doesn't optimize for the 99% of use cases but rather the one single time where users will be running into this. It's probably way more popular that the users will run into server quotas (which is obviously not checked in Terraform).

**Comment by catsby** [comment]
>documented limits in our API, don't you think it's better that our API side checks this?

I'm not sure I follow. Do you mean the validation I proposed? Ideally we would catch any known unacceptable values at Plan time, and not at Run time.

**Comment by davidhalter** [comment]
@catsby I mean the validation you proposed.

Hmm, I can see where you're coming from. It just adds a certain inflexibility to terraform. If we e.g. allow a single user to add IP /48 networks, it's clear that the documentation will still state that it's only possible to use `/56` networks. This is the default for all other users.

If terraform catches this, they will not really be able to use terraform. We already have users using their own /28 IPv4 networks, I would be sad to see this limit.

**Comment by thetonymaster** [comment]
@davidhalter in that case, what would be a validation that future-proofs this issue? checking for a valid mask? or a series of masks?

**Comment by davidhalter** [comment]
@thetonymaster Well, let's do the rest of the changes and see what @catsby thinks. I'm all for not validating the prefix, because it hurts us in the long run. However since @catsby has the final say I'm just waiting for his decision.

**Comment by thetonymaster** [comment]
@catsby I set the default at 10 for `volume_size_gb`, however if I try to add a the `computed` attribute, tests fail.

**Comment by catsby** [comment]
Regarding the validation, the validation suggestion is aimed at avoiding _run time_ validation errors that could be caught at _plan time_, if possible. I don't have the context you both seem to have, so I will defer to your judgement. I don't consider it a blocker 😄 , so I'll go ahead and strike it. 

For `volume_size_gb`, I may have been trying to guard against a situation with removing optional attributes, but as this one is `ForceNew`, it is likely not necessary. I'll strike that one as well.

Regarding `ssh_keys`:

- I believe this method will allow the use of `file()` builtin method so users don't need to statically copy/paste the public key here, so that's good
- with `ssh_host_keys` and `ssh_fingerprints`, is there any reconciliation between these and `ssh_keys`? If a user were to add/remove a key out-of-band of Terraform, how would Terraform alert the operator that the configuration doesn't match reality (drift detection)? From looking at the API I assume that `ssh_keys` and `ssh_host_keys` represent the same thing(s), so I would think to not include `ssh_host_keys` at all, and read that value into `ssh_keys`. Assuming that is correct, I believe you would want `ssh_keys` to be a `TypeSet` and not `TypeList`, as List will require both config and server response to be in the same order, where as Set would not. Please let me know if any of those assumptions are incorrect 😄 

Let me know if you have other questions!

**Comment by davidhalter** [comment]
> ssh_keys and ssh_host_keys  represent the same thing

No. They are entirely different things. This might not be clear enough in the documentation, but host keys are the *host* public keys that were calculated there whereas `ssh_keys` is simply the users public key. `ssh_fingerprints` are basically the calculated fingerprints of the `ssh_host_keys`.

> Let me know if you have other questions!

There's a pull request open now for the open issues here: #13. Can you please check it quickly and see if that's what you want to be changed?

> ssh_keys are not read from the API. Are they available? Can you describe the expected behavior of this attribute?

This is something "writeonly", i.e. you can really not read it from the API.

------

IMO all the issues above have now either been discussed in this issue or addressed in the pull request.

**Comment by catsby** [comment]
Thanks for clarifying! I've reviewed #13 and besides one `return` line change, I think we're good here!

**Comment by catsby** [comment]
I did open #14 that should be looked at as well
